Plugins/GitUIPluginInterfaces/ISettingsSource.cs

```diff
@@ -165,15 +183,126 @@ public Color GetColor(string name, Color defaultValue)
             });
         }
 
+        [Obsolete("Use new SetValue<T> method.")]
         public void SetString(string name, string? value)
         {
             SetValue(name, value, s => string.IsNullOrEmpty(s) ? null : s);
         }
 
+        [Obsolete("Use new GetValue<T> method.")]
         [return: NotNullIfNotNull("defaultValue")]
         public string? GetString(string name, string? defaultValue)
         {
             return GetValue(name, defaultValue, x => x);
         }
+
+#pragma warning disable CS0618 // Type or member is obsolete
+        public T GetValue<T>(string key, T defaultValue = default!)
```

Comment:
Why move to an interface that loses type information (pushing it to `T`) then introduce dispatch on the value of `T`?

Why not just move `GetBool`, `GetInt`, ... to the interface?

Reply:
> Why move to an interface that loses type information (pushing it to T) then introduce dispatch on the value of T?

As far as I understand, we are not losing information. I am planning further steps and the current implementation will be changed.

> Why not just move GetBool, GetInt, ... to the interface?

After adding new types, we will have to change the interface. In the current case, we only add the implementation.
As I understand it, the generic types are for this.

Reply:
> As far as I understand, we are not losing information

I was referring to runtime information. The `switch` over `typeof(T)` will happen at runtime and has a cost associated with it. This cost did not exist previously.

Is there a benefit to changing this that I don't see?

Reply:
I understand that this adds transformations that were not there, but this is temporary.
This is required to work with the current implementation.
I announced further thoughts in #9032.

Reply:
> Why move to an interface that loses type information (pushing it to `T`) then introduce dispatch on the value of `T`?
> 
> Why not just move `GetBool`, `GetInt`, ... to the interface?

I personally find `GetBool` et al. rather distasteful, and an unnecessary cognitive load. All changes outside this class look great - the public API surface is simple and clean. The changes here do look ugly and may result in additional boxing.

We should be able to provide additional overloads for value-types, using `struct` and `class` constraints. 🤔

Reply:
> I personally find GetBool et al. rather distasteful, and an unnecessary cognitive load.

Can you elaborate? This pattern is not common in the BCL, for example. Also accepting `T` means you might expect `GetValue<MyCustomType>` to work, but it fails at runtime instead of compile time.

Reply:
> Also accepting T means you might expect GetValue<MyCustomType> to work, but it fails at runtime instead of compile time.

Are you suggesting returning null instead of throwing an exception?

Reply:
No, I'm suggesting that the API shouldn't allow using invalid types.

Look at the methods on [`BinaryReader`](https://docs.microsoft.com/en-us/dotnet/api/system.io.binaryreader?view=net-5.0#methods) for example. It has a method per type.

[`BinaryWriter`](https://docs.microsoft.com/en-us/dotnet/api/system.io.binarywriter?view=net-5.0#methods) does as well, though they're overloads with the same name.

Not just in the BCL either. Check out JSON.NET's [`JsonReader`](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_JsonReader.htm) class too.

Reply:
Thank. I'll see.
But this interface will work with all types. This was not required to integrate with the current implementation, so I did not extend it. Since I am planning further steps.

Reply:
Will it actually be extensible? All types would have to be in that switch statement, no?

Reply:
If we talk about the current implementation, then yes. For objects, we could add.

`JsonConvert.DeserializeObject<T>(value);`

or from Microsoft

`JsonSerializer.Serialize<T>(value);`

In any case, consumers are implementation independent.

Reply:
> Look at the methods on BinaryReader for example. It has a method per type.
>
>BinaryWriter does as well, though they're overloads with the same name.
>
>Not just in the BCL either. Check out JSON.NET's JsonReader class too.

BinaryReader and BinaryWriter.
I think they are done this way, because they work only for these types.

JsonReader.
Because the JSON format has these types in the specification or so.

Reply:
Please see last commit. I think this will be more clear interface.

Reply:
Did such a test.

![image](https://user-images.githubusercontent.com/3169265/112732933-a34a4100-8f45-11eb-9ba4-e284f4aa7a18.png)

![image](https://user-images.githubusercontent.com/3169265/112732958-d260b280-8f45-11eb-8d2a-2ff2c6d6dfb5.png)

https://github.com/dotnet/runtime/blob/7795971839be34099b07595fdcf47b95f048a730/src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs#L124-L147

I don't know how it works correctly, but null returns. Perhaps we shouldn't complicate things too? 😄

Reply:
> Look at the methods on [`BinaryReader`](https://docs.microsoft.com/en-us/dotnet/api/system.io.binaryreader?view=net-5.0#methods) for example. It has a method per type.
> 
> [`BinaryWriter`](https://docs.microsoft.com/en-us/dotnet/api/system.io.binarywriter?view=net-5.0#methods) does as well, though they're overloads with the same name.
> 
> Not just in the BCL either. Check out JSON.NET's [`JsonReader`](https://www.newtonsoft.com/json/help/html/T_Newtonsoft_Json_JsonReader.htm) class too.

This is a good point. I didn't consider these.

Reply:
> [dotnet/runtime@`7795971`/src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs#L124-L147](https://github.com/dotnet/runtime/blob/7795971839be34099b07595fdcf47b95f048a730/src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs#L124-L147)
> 
> I don't know how it works correctly, but null returns. Perhaps we shouldn't complicate things too? 😄

To use this I'd guess we'd need to provide same kind of plumbing ASP.NET Core uses... It's been awhile since I used ASP.NET Core last ime, and haven't really looked in how it plumbs configs.

I think it would be prudent of us to retain `ReadXyz` methods, where `Xyz` denotes a known type, and have a single `Write(Xyz)` method with required overloads.

Reply:
Consider whether extension methods could play a role here too. It looks like the underlying data is `string`. If that's the case, the interface could work only with `string` and all the typed get/set methods could be extensions.